ARCore/app/src/main/java/com/novoda/spikes/arcore/MainActivity.kt

```diff
@@ -0,0 +1,87 @@
+package com.novoda.spikes.arcore
+
+import android.opengl.GLSurfaceView
+import android.os.Bundle
+import android.support.v7.app.AppCompatActivity
+import android.widget.Toast
+import com.google.ar.core.Session
+import com.google.ar.core.exceptions.CameraNotAvailableException
+import com.novoda.spikes.arcore.google.helper.TapHelper
+import com.novoda.spikes.arcore.helper.ARCoreDependenciesHelper
+import com.novoda.spikes.arcore.helper.CameraPermissionHelper
+import com.novoda.spikes.arcore.rendering.NovodaSurfaceViewRenderer
+import kotlinx.android.synthetic.main.activity_main.*
+
+class MainActivity : AppCompatActivity() {
+    private val arSession: Session by lazy {
+        Session(this).apply {
+            renderer.setSession(this)
+        }
+    }
+
+    private val renderer: NovodaSurfaceViewRenderer by lazy {
+        NovodaSurfaceViewRenderer(this, debugViewDisplayer, tapHelper)
+    }
+    private val debugViewDisplayer: DebugViewDisplayer by lazy {
+        DebugViewDisplayer(debugTextView)
+    }
+    private val tapHelper: TapHelper by lazy {
+        TapHelper(this)
+    }
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+        setContentView(R.layout.activity_main)
+        setupSurfaceView()
+    }
+
+    private fun setupSurfaceView() {
+        surfaceView.preserveEGLContextOnPause = true
+        surfaceView.setEGLContextClientVersion(2)
+        surfaceView.setEGLConfigChooser(8, 8, 8, 8, 16, 0) // Alpha used for plane blending.
+        surfaceView.setRenderer(renderer)
+        surfaceView.renderMode = GLSurfaceView.RENDERMODE_CONTINUOUSLY
+        surfaceView.setOnTouchListener(tapHelper)
+    }
+
+    override fun onResume() {
+        super.onResume()
+        checkAREnvironment()
+    }
+
+    private fun checkAREnvironment() {
+        ARCoreDependenciesHelper.isARCoreIsInstalled(this).apply {
+            if (this is ARCoreDependenciesHelper.Result.Failure) {
+                showMessage(message)
+            } else if (CameraPermissionHelper.isCameraPermissionGranted(this@MainActivity)) {
+                createOrResumeARSession()
+            }
```

Comment:
With the introduction of a sealed class to model `Result` this is not looking so good ☝️  - Is there a better way to do that?

Reply:
hmmm usually with `sealed` classes people tend to replace `if` with `when`
this code should almost compile 😅 : (also using static imports)

```kotlin
ARCoreDependenciesHelper.isARCoreIsInstalled(this).apply {
  when (this) {
    is Failure -> showMessage(message)
    is Success && CameraPermissionHelper.isCameraPermissionGranted(this@MainActivity) -> createOrResumeARSession()
  }
}
```